[quant][graphmode][fx] Merge all quantization mode

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45539 [quant][graphmode][fx][eagermode] Support sigmoid/hardsigmoid/tanh in eager and fx graph mode
* #45538 [quant] Add FixedQParamsFakeQuantize module
* #45473 [quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input
* **#45292 [quant][graphmode][fx] Merge all quantization mode**
* #45343 [quant] Use PlaceholderObserver as default dynamic quant observer

Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

jerryzh168 committed:
[quant][graphmode][fx] Merge all quantization mode

Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/utils.py:
```diff
@@ -138,3 +138,38 @@ def get_next_qparams_idx(module, qparams):
         inputs.append(graph.create_node('get_attr', qparam_full_path))
     return graph.create_node('call_function', quantize_op, tuple(inputs), {})
+
+def activation_is_dynamically_quantized(qconfig):
```
looks super clean, thanks!

vkuzo reviewed (approved):
looks awesome, the new names are much easier to read!

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/utils.py:
```diff
@@ -138,3 +138,38 @@ def get_next_qparams_idx(module, qparams):
+    return weight.dtype
+
+def weight_is_quantized(qconfig):
+    """ Given a qconfig, decide if the activation needs to be
```
doc string needs to be updated for this one

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -308,11 +310,10 @@ def get_qconfig(module_name):
-            self.patterns = get_dynamic_quant_patterns()
-        else:
-            self.patterns = get_quant_patterns()
+        # if is_dynamic_quant:
+        #     self.patterns = get_dynamic_quant_patterns()
+        # else:
```
Can we remove the comment block here?

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -404,8 +405,9 @@ def insert_observer(node, observer, device):
-                if self.is_dynamic_quant:
+                # don't need to insert observer for output if activation does not
+                # need to be statically quantized
+                if not activation_is_statically_quantized(qconfig):
```
Can we use `activation_is_dynamically_quantized` instead here for better readability?

jerryzh168 replied:
I'm planning to remove that check since it checks for both dynamic and weight only dtypes.

supriyar replied:
Not sure I follow. What is the check we will use for dynamic quant in that case?

jerryzh168 replied:
activation_is_dynamically_quantized checks for float32 dtype which can be both dynamic or weight only quantization.

supriyar commented:
Will we also be updating the top level APIs to remove the `quantize_{static, dynamic}_fx` calls ? https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize_fx.py#L108-L218

jerryzh168 commented:
> Will we also be updating the top level APIs to remove the `quantize_{static, dynamic}_fx` calls ? https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize_fx.py#L108-L218

yes will do that, I'm waiting for previous PRs to be landed to avoid conflicts.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Merge all quantization mode"


Summary:
This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23913105](https://our.internmc.facebook.com/intern/diff/D23913105)

[ghstack-poisoned]

t-vi commented:
This seems to break the CI on ROCm with an error that looks like it is this patch:
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.7-py3.6-test2/5653//console
```
FAIL: test_qat_prepare_device_affinity (quantization.test_quantize_fx.TestQuantizeFx)
06:50:29 ----------------------------------------------------------------------
06:50:29 Traceback (most recent call last):
06:50:29   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_quantized.py", line 122, in test_fn
06:50:29     qfunction(*args, **kwargs)
06:50:29   File "/var/lib/jenkins/workspace/test/quantization/test_quantize_fx.py", line 268, in test_qat_prepare_device_affinity
06:50:29     model = prepare_fx(model, qconfig_dict)
06:50:29   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/quantization/quantize_fx.py", line 127, in prepare_fx
06:50:29     'eval mode'
06:50:29 AssertionError: prepare_fx only works for models ineval mode
06:50:29 
```

mruberry commented:
Unlanding. @t-vi nailed it.
